test_maxnrows_on_large_dataset is unstable

As seen on Jenkins:
```
________________________ test_maxnrows_on_large_dataset ________________________

    def test_maxnrows_on_large_dataset():
        # Limit number of threads during reading, otherwise the test may fail on a
        # machine with many cores (all chunks read at once, in the same amount of
        # time as the single chunk containing the first 5 rows).
        src = "A,B,C\n" + "\n".join("%06d,x,1" % i for i in range(2000000))
        t0 = time.time()
        d0 = dt.fread(src, nthreads=4, max_nrows=5, verbose=True)
        t0 = time.time() - t0
        d0.internal.check()
        assert d0.shape == (5, 3)
        assert d0.topython() == [[0, 1, 2, 3, 4], ["x"] * 5, [True] * 5]
        t1 = time.time()
        d1 = dt.fread(src, nthreads=4)
        t1 = time.time() - t1
        assert d1.shape == (2000000, 3)
>       assert t0 < t1 / 2, ("Reading with max_nrows=5 should be faster than "
                             "reading the whole dataset")
E       AssertionError: Reading with max_nrows=5 should be faster than reading the whole dataset
E       assert 0.09199047088623047 < (0.13933181762695312 / 2)
```

<details><summary>The verbose output from `fread`:</summary>

```
[1] Prepare for reading
  Input is a string of length 23000005, treating it as raw text
  Using 4 threads (requested=4, max.available=40)
  max_nrows = 5
  sep = <auto-detect>
  Decimal separator = '.'
  Quote char = (")
  na_strings = ["NA"]
  strip_whitespace = True
  skip_blank_lines = False
  ==== file sample ====
  A,B,C
  000000,x,1
  000001,x,1
  000002,x,1
  000003,x,1
  =====================
  LF character (\n) found in input, \r-only newlines will not be recognized
[2] Detect separator, quoting rule, and ncolumns
  sep=',' with 100 lines of 3 fields using quote rule 0
  Detected 3 columns
  sep = ','
  Quote rule = 0
[3] Detect column types and header
  Number of threads reduced to 1 because due to max_nrows=5 we estimate the amount of data to be read will be small
  The input will be read in 2 chunks of size 8 each
  Number of sampling jump points = 1 because max_nrows parameter is used
  Type codes (jump 000): isb
  `header` determined to be True due to column 1 containing a string on row 1 and type Int32 in the rest of the sample.
  =====
  Sampled 6 rows (handled \n inside quoted fields) at 1 jump point(s)
  Bytes from first data row to the end of last row: 23000005
  Line length: mean=11.00 sd=0.00 min=11 max=11
  Estimated number of rows: 23000005 / 11.00 = 2090910
  Initial alloc = 2300001 rows (2090910 + 10%) using bytes/max(mean-2*sd,min) clamped between [1.1*estn, 2.0*estn]
  Alloc limited to nrows=5 according to the provided max_nrows argument.
[4] Assign column names
[5] Apply user overrides on column types
  After 0 type and 0 drop user overrides : isb
  Allocating 3 column slots with 5 rows
[6] Read the data
  Number of threads reduced to 1 because due to max_nrows=5 we estimate the amount of data to be read will be small
  The input will be read in 2 chunks of size 83 each
  Too few rows allocated, reallocating to 5 rows
[7] Finalize the datatable
  =============================
  Read 5 rows x 3 columns from 21.93MB input in 00:00.091s
   = 0.091s (100%) memory-mapping input file
   + 0.000s ( 0%) detecting parse parameters
   + 0.000s ( 0%) detecting column types using 6 sample rows
   + 0.000s ( 0%) allocating [5 x 3] frame (444 bytes) of which 5 (100%) rows used
   + 0.000s ( 0%) reading data using 1 thread
      = 0.000s ( 0%) reading into row-major buffers
      + 0.000s ( 0%) saving into the output frame
      + 0.000s ( 0%) waiting
   + 0.000s ( 0%) creating the final Frame
```
</details>

--- a/c/csv/fread.cc
+++ b/c/csv/fread.cc
@@ -90,8 +90,10 @@ std::unique_ptr<DataTable> FreadReader::read()
       }
     }
     if (verbose) {
-      trace("After %d type and %d drop user overrides : %s",
-            nUserBumped, ndropped, columns.printTypes());
+      if (nUserBumped || ndropped) {
+        trace("After %d type and %d drop user overrides : %s",
+              nUserBumped, ndropped, columns.printTypes());
+      }
       trace("Allocating %d column slots with %zd rows",
             ncols - ndropped, allocnrow);
     }

--- a/c/csv/fread.h
+++ b/c/csv/fread.h
@@ -18,6 +18,10 @@
 extern const long double pow10lookup[701];
 extern const uint8_t hexdigits[256];
 extern const uint8_t allowedseps[128];
+namespace dt {
+namespace read {
+  struct ChunkCoordinates;
+}}
 
 
 struct FreadTokenizer {
@@ -69,6 +73,11 @@ struct FreadTokenizer {
   int countfields();
   bool skip_eol();
   bool at_eof() const { return ch == eof; }
+
+  bool next_good_line_start(
+    const dt::read::ChunkCoordinates& cc, int ncols, bool fill, bool skipEmptyLines);
+
+
 };
 
 

--- a/c/csv/reader.cc
+++ b/c/csv/reader.cc
@@ -514,6 +514,7 @@ void GenericReader::open_input() {
   } else if ((text = text_arg.as_cstring(&size))) {
     input_mbuf = MemoryRange::external(text, size + 1);
     extra_byte = 1;
+    input_is_string = true;
 
   } else if ((filename = file_arg.as_cstring())) {
     input_mbuf = MemoryRange::overmap(filename, /* extra = */ 1);

--- a/c/csv/reader.h
+++ b/c/csv/reader.h
@@ -86,7 +86,8 @@ class GenericReader
     size_t line;
     int32_t fileno;
     bool cr_is_newline;
-    int : 24;
+    bool input_is_string{ false };
+    int : 16;
     dt::read::Columns columns;
     double t_open_input{ 0 };
 

--- a/c/csv/reader_fread.cc
+++ b/c/csv/reader_fread.cc
@@ -384,14 +384,13 @@ void FreadReader::detect_sep_and_qr() {
 class ColumnTypeDetectionChunkster {
   public:
     const FreadReader& f;
-    dt::read::FreadParallelReader fcr;
     FreadTokenizer fctx;
     size_t nchunks;
     size_t chunk_distance;
     const char* last_row_end;
 
     ColumnTypeDetectionChunkster(FreadReader& fr, FreadTokenizer& ft)
-    : f(fr), fcr(fr, nullptr), fctx(ft)
+        : f(fr), fctx(ft)
     {
       nchunks = 0;
       chunk_distance = 0;
@@ -447,7 +446,8 @@ class ColumnTypeDetectionChunkster {
         while (*tch == '\n' || *tch == '\r') tch++;
 
         if (tch < f.eof) {
-          bool ok = fcr.next_good_line_start(cc, fctx);
+          bool ok = fctx.next_good_line_start(cc, static_cast<int>(f.get_ncols()),
+                                              f.fill, f.skip_blank_lines);
           if (ok) cc.start = fctx.ch;
         }
       }
@@ -563,7 +563,7 @@ void FreadReader::detect_column_types()
   double sumLenSq = 0.0;
   int minLen = INT32_MAX;   // int_max so the first if(thisLen<minLen) is always true; similarly for max
   int maxLen = -1;
-  int rows_to_sample = static_cast<int>(std::min<size_t>(max_nrows, 99)) + 1;
+  int rows_to_sample = static_cast<int>(std::min<size_t>(max_nrows, 100));
 
   // Start with all columns having the smallest possible type
   columns.setType(PT::Mu);
@@ -1092,6 +1092,42 @@ int FreadTokenizer::countfields()
 }
 
 
+// Find the next "good line", in the sense that we find at least 5 lines
+// with `ncols` fields from that point on.
+bool FreadTokenizer::next_good_line_start(
+  const dt::read::ChunkCoordinates& cc, int ncols, bool fill, bool skipEmptyLines)
+{
+  // int ncols = static_cast<int>(f.get_ncols());
+  // bool fill = f.fill;
+  // bool skipEmptyLines = f.skip_blank_lines;
+  ch = cc.start;
+  const char* end = cc.end;
+  int attempts = 0;
+  while (ch < end && attempts++ < 10) {
+    while (ch < end && *ch != '\n' && *ch != '\r') ch++;
+    if (ch == end) break;
+    skip_eol();  // updates `ch`
+    // countfields() below moves the parse location, so store it in `ch1` in
+    // order to revert to the current parsing location later.
+    const char* ch1 = ch;
+    int i = 0;
+    for (; i < 5; ++i) {
+      // `countfields()` advances `ch` to the beginning of the next line
+      int n = countfields();
+      if (n != ncols &&
+          !(ncols == 1 && n == 0) &&
+          !(skipEmptyLines && n == 0) &&
+          !(fill && n < ncols)) break;
+    }
+    ch = ch1;
+    // `i` is the count of consecutive consistent rows
+    if (i == 5) return true;
+  }
+  return false;
+}
+
+
+
 
 //==============================================================================
 // FreadObserver
@@ -1154,8 +1190,10 @@ void FreadObserver::report() {
           humanize_number(n_cols_read), (n_cols_read == 1 ? "" : "s"),
           filesize_to_str(input_size),
           total_minutes, total_seconds);
-  g.trace(" = %*.3fs (%2.0f%%) memory-mapping input file", p,
-          g.t_open_input, 100 * g.t_open_input / total_time);
+  g.trace(" = %*.3fs (%2.0f%%) %s", p,
+          g.t_open_input, 100 * g.t_open_input / total_time,
+          g.input_is_string? "converting input string into bytes"
+                           : "memory-mapping input line");
   g.trace(" + %*.3fs (%2.0f%%) detecting parse parameters", p,
           params_time, 100 * params_time / total_time);
   g.trace(" + %*.3fs (%2.0f%%) detecting column types using %s sample rows", p,

--- a/c/read/fread/fread_parallel_reader.cc
+++ b/c/read/fread/fread_parallel_reader.cc
@@ -33,43 +33,6 @@ std::unique_ptr<ThreadContext> FreadParallelReader::init_thread_context() {
 }
 
 
-// Find the next "good line", in the sense that we find at least 5 lines
-// with `ncols` fields from that point on.
-bool FreadParallelReader::next_good_line_start(
-  const ChunkCoordinates& cc, FreadTokenizer& tokenizer) const
-{
-  int ncols = static_cast<int>(f.get_ncols());
-  bool fill = f.fill;
-  bool skipEmptyLines = f.skip_blank_lines;
-  const char*& ch = tokenizer.ch;
-  ch = cc.start;
-  const char* eof = cc.end;
-  int attempts = 0;
-  while (ch < eof && attempts++ < 10) {
-    while (ch < eof && *ch != '\n' && *ch != '\r') ch++;
-    if (ch == eof) break;
-    tokenizer.skip_eol();  // updates `ch`
-    // countfields() below moves the parse location, so store it in `ch1` in
-    // order to revert to the current parsing location later.
-    const char* ch1 = ch;
-    int i = 0;
-    for (; i < 5; ++i) {
-      // `countfields()` advances `ch` to the beginning of the next line
-      int n = tokenizer.countfields();
-      if (n != ncols &&
-          !(ncols == 1 && n == 0) &&
-          !(skipEmptyLines && n == 0) &&
-          !(fill && n < ncols)) break;
-    }
-    ch = ch1;
-    // `i` is the count of consecutive consistent rows
-    if (i == 5) return true;
-  }
-  return false;
-}
-
-
-
 void FreadParallelReader::adjust_chunk_coordinates(
   ChunkCoordinates& cc, ThreadContext* ctx) const
 {
@@ -80,7 +43,8 @@ void FreadParallelReader::adjust_chunk_coordinates(
     const char* start = cc.start;
     while (*start=='\n' || *start=='\r') start++;
     cc.start = start;
-    if (next_good_line_start(cc, fctx->tokenizer)) {
+    int ncols = static_cast<int>(f.get_ncols());
+    if (fctx->tokenizer.next_good_line_start(cc, ncols, f.fill, f.skip_blank_lines)) {
       cc.start = fctx->tokenizer.ch;
     }
   }

--- a/c/read/fread/fread_parallel_reader.h
+++ b/c/read/fread/fread_parallel_reader.h
@@ -32,9 +32,6 @@ class FreadParallelReader : public ParallelReader {
 
     virtual void read_all() override;
 
-    bool next_good_line_start(
-      const ChunkCoordinates& cc, FreadTokenizer& tokenizer) const;
-
   protected:
     virtual std::unique_ptr<ThreadContext> init_thread_context() override;
 

--- a/c/read/parallel_reader.cc
+++ b/c/read/parallel_reader.cc
@@ -57,7 +57,7 @@ void ParallelReader::determine_chunking_strategy() {
     if (inputSize_reduced) {
       // If chunkCount is 1 then we'll attempt to read the whole input
       // with the first chunk, which is not what we want...
-      chunkCount++;
+      chunkCount += 2;
       g.trace("Number of threads reduced to %d because due to max_nrows=%zu "
               "we estimate the amount of data to be read will be small",
               nthreads, nrows_max);
@@ -263,6 +263,9 @@ void ParallelReader::read_all()
 
 void ParallelReader::realloc_output_columns(size_t ichunk, size_t new_alloc)
 {
+  if (new_alloc == nrows_allocated) {
+    return;
+  }
   if (ichunk == chunkCount - 1) {
     // If we're on the last jump, then `new_alloc` is exactly how many rows
     // will be needed.

--- a/tests/fread/test_fread_api.py
+++ b/tests/fread/test_fread_api.py
@@ -805,15 +805,19 @@ def test_skip_to_string_bad():
 # `max_nrows`
 #-------------------------------------------------------------------------------
 
-def test_fread_max_nrows():
+def test_fread_max_nrows(capsys):
     d0 = dt.fread("A,B,C\n"
                   "1,foo,1\n"
                   "3,bar,0\n"
                   "5,baz,0\n"
-                  "7,meh,1\n", max_nrows=2)
+                  "7,meh,1\n", max_nrows=2, verbose=True)
+    out, err = capsys.readouterr()
     d0.internal.check()
     assert d0.names == ("A", "B", "C")
     assert d0.topython() == [[1, 3], ["foo", "bar"], [True, False]]
+    assert "Allocating 3 column slots with 2 rows" in out
+    assert "Too few rows allocated" not in out
+    assert "converting input string into bytes" in out
 
 
 def test_fread_max_nrows_0rows():
@@ -821,6 +825,20 @@ def test_fread_max_nrows_0rows():
     d0.internal.check()
     assert d0.names == ("A", )
     assert d0.shape == (0, 1)
+
+
+@pytest.mark.xfail()
+def test_fread_max_nrows_correct_types():
+    d0 = dt.fread("A,B\n"
+                  "1,2\n"
+                  "3,4\n"
+                  "5,6\n"
+                  "foo,bar\n"
+                  "zzzzz\n", max_nrows=3)
+    assert d0.names == ("A", "B")
+    assert d0.shape == (3, 2)
+    assert d0.ltypes == (ltype.int, ltype.int)
+    assert d0.topython() == [[1, 3, 5], [2, 4, 6]]
 
 
 

--- a/tests/fread/test_fread_small.py
+++ b/tests/fread/test_fread_small.py
@@ -907,7 +907,7 @@ def test_int64s_and_typebumps(capsys):
            ]
     src[2][111] = 4568034971487098  # i64-2s
     src[3][111] += 0.11  # f64-1
-    src[4][111] = 11111111111111.11  #  f64-2
+    src[4][111] = 11111111111111.11  # f64-2
     src[6][111] = "111e"  # s32-2
     src[7][111] = "1111111111111111111111"  # s32-3
     src[8][111] = "1.23e"
@@ -1004,7 +1004,7 @@ def test_maxnrows_on_large_dataset():
     # Limit number of threads during reading, otherwise the test may fail on a
     # machine with many cores (all chunks read at once, in the same amount of
     # time as the single chunk containing the first 5 rows).
-    src = "A,B,C\n" + "\n".join("%06d,x,1" % i for i in range(2000000))
+    src = b"A,B,C\n" + b"\n".join(b"%06d,x,1" % i for i in range(2000000))
     t0 = time.time()
     d0 = dt.fread(src, nthreads=4, max_nrows=5, verbose=True)
     t0 = time.time() - t0
@@ -1012,7 +1012,7 @@ def test_maxnrows_on_large_dataset():
     assert d0.shape == (5, 3)
     assert d0.topython() == [[0, 1, 2, 3, 4], ["x"] * 5, [True] * 5]
     t1 = time.time()
-    d1 = dt.fread(src, nthreads=4)
+    d1 = dt.fread(src, nthreads=4, verbose=True)
     t1 = time.time() - t1
     assert d1.shape == (2000000, 3)
     assert t0 < t1 / 2, ("Reading with max_nrows=5 should be faster than "

